### DISCUSSION TITLE
Accessing response extensions

### DISCUSSION THREAD

**Discussion by darrell-roberts** [tokio-rs/axum#776] [Q&A] [answered]
Hi,

I can use `AddExtensionLayer` with my routes and access the type from a custom extractor and of course the handler functions. I was under the impression that the extensions() from the `Response` would also have the same extensions accessible to the `Request`. It seems that this is not the case. 

For example:
```rust
Router::new()
    .route(
      "/user/:id",
      get(handlers::get_user).layer(AndThenLayer::new(middleware::hash_user)),
    )
...
  .layer(AddExtensionLayer::new(Arc::new(app_config)))
```
middleware.rs
```rust
fn get_hash_prefix(response: &Response<BoxBody>) -> &str {
  response
    .extensions()
    .get::<Arc<AppConfig>>()
    .map(|config| config.hash_prefix.as_str())
    .unwrap_or("default_prefix")
}

pub async fn hash_user(
  response: Response<BoxBody>,
) -> Result<Response<BoxBody>, Infallible> {
  let prefix = get_hash_prefix(&response).to_owned();

  event!(target: FRAMEWORK_TARGET, Level::DEBUG, "Using hash prefix: {prefix}");

  if response.status().is_success() {
    let result = response.map(move |body| {
      boxed(body.map_data(move |bytes| {
        match serde_json::from_slice(&bytes)
          .map(|b: User| b.hash(&prefix))
          .ok()
        {
          Some(hashed) => Bytes::from(serde_json::to_vec(&hashed).unwrap()),
          None => bytes,
        }
      }))
    });
    Ok(result)
  } else {
    Ok(response)
  }
}
```
The `get_hash_prefix` always returns the "default_prefix". The `Arc<AppConfig>` does not appear to be stored in the `Response` extensions.

**Comment by davidpdrsn** [comment, marked as answer]
Extensions on the request aren't automatically copied to the response. You have to do that manually for the necessary extensions.

**Comment by darrell-roberts** [comment]
Ok. That's what I thought. In this case I opted to create middleware via `layer_fn`. This way I can source the extensions from the request. It's more code then using the `AndThenLayer` but it works.
